king/king.wdl

```diff
@@ -0,0 +1,419 @@
+		cpu: cpu
+		docker: "uwgac/topmed-master@sha256:0bb7f98d6b9182d4e4a6b82c98c04a244d766707875ddfd8a48005a9f5c5481e"
+		disks: "local-disk " + final_disk_dize + " HDD"
+		memory: "${memory} GB"
+		preemptibles: "${preempt}"
+	}
+	output {
+		File king_matrix = glob("*.RData")[0]
+		File config_file = "king_to_matrix.config"
+	}
+
+}
+
+# [5] kinship_plots -- 
+
+task kinship_plots {
+	input {
+		File kinship_file
+		String? kinship_method
+
+		# optional
+		Float? kinship_plot_threshold = 0.04419417382
+		File? phenotype_file
+		String? group
+		File? sample_include_file
+		String? out_prefix = "kinship"
+		
+		# runtime attributes
+		Int addldisk = 5
+		Int cpu = 2
+		Int memory = 4
+		Int preempt = 3
+	}
+
+	# Estimate disk size required
+	Int kinship_size = ceil(size(kinship_file, "GB"))
+	Int final_disk_dize = kinship_size + addldisk
+
+	# Workaround for optional files
+	Boolean defPhenotypeFile = defined(phenotype_file)
+	Boolean defSampleInclude = defined(sample_include_file)
+
+	
+	command {
+		set -eux -o pipefail
+
+		echo "Generating config file"
+		# noquotes = argument has no quotes in the config file to match CWL config EXACTLY
+		python << CODE
+		import os
+
+		f = open("kinship_plots.config", "a")
+		f.write('kinship_file "~{kinship_file}"\n')
+
+		if "~{kinship_method}" != "":
+			if "~{kinship_method}" == "king_robust":
+				f.write('kinship_method king\n')
+			else:
+				f.write('kinship_method "~{kinship_method}"\n')
+		else:
+			f.write('kinship_method "king_ibdseg"\n')
+
+		# check for empty string
+		if "~{kinship_plot_threshold}" != "":
+				f.write('kinship_plot_threshold "~{kinship_plot_threshold}"\n')		
+
+		if "~{out_prefix}" != "":
+			f.write('out_file_all "' + "~{out_prefix}" + '_all.pdf"\n')
+			f.write('out_file_cross "' + "~{out_prefix}" + '_cross_group.pdf"\n')
+			f.write('out_file_study "' + "~{out_prefix}" + '_within_group.pdf"\n')
+
+		if "~{defPhenotypeFile}" == "true":
+			f.write('phenotype_file ""~{phenotype_file}""\n')
+
+		if "~{group}" != "":
+			f.write('study ""~{group}""\n')		
+
+		if "~{defSampleInclude}" == "true":
+			f.write('sample_include_file ""~{sample_include_file}""\n')
+
+
+		f.close()
+		CODE
+
+		echo "Calling R script kinship_plots.R"
+		Rscript /usr/local/analysis_pipeline/R/kinship_plots.R kinship_plots.config
+	}
+
+	runtime {
+		cpu: cpu
+		docker: "uwgac/topmed-master@sha256:0bb7f98d6b9182d4e4a6b82c98c04a244d766707875ddfd8a48005a9f5c5481e"
+		disks: "local-disk " + final_disk_dize + " HDD"
+		memory: "${memory} GB"
+		preemptibles: "${preempt}"
+	}
+	output {
+		File kinship_plots = glob("*.pdf")[0]
+		File config_file = "kinship_plots.config"
+	}
+
+}
+
+workflow king {
+	input {
+		File gds_file
+		File? sample_include_file
+		File? variant_include_file
+		String? out_prefix
+		File? phenotype_file
+		Float? kinship_plot_threshold
+		String? group
+		Float? sparse_threshold
+		String? kinship_method # not included in king-ibdseg-wf.cwl but is an input in king_to_matrix task
```

Comment:
When this sort of thing happens, you actually don't have to define it as a workflow input. You can leave it as an input in the task's input section only and simply pretend it doesn't exist in the workflow section. On the Terra UI, it will still show up as a task-specific input. This will not only more closely mimic the cwl, but will also make it slightly clearer in Terra's UI.

Reply:
Makes sense, done!!